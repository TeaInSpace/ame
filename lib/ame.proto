syntax = "proto3";

package ame.v1 ;

message TaskIdentifier {
  string name = 1;
}

enum TaskType {
  Pipenv = 0;
  Mlflow = 1;
  Poetry = 2;
}

message Empty {
}

message ProjectFileIdentifier {
  string taskid= 1;
  string filepath = 2;
}

message FileChunk {
  bytes contents = 1;
}

message TaskProjectDirectoryStructure {
  string projectid = 1;
  TaskIdentifier taskid = 2; 
  repeated string paths = 3;
}

message ProjectFileChunk {
  oneof messages {
    FileChunk chunk = 1;
    ProjectFileIdentifier identifier = 2;
  }
}

message TaskLogRequest {
  TaskIdentifier taskid = 1;
  optional int32 start_from = 2;
  optional bool watch = 3;
}

message LogEntry {
  bytes contents = 1;
}

message ProjectSourceCfg {
  optional GitProjectSource git = 1;
}

message GitProjectSource {
  string repository = 1;
  optional string username = 2;
  optional string secret = 3;
  optional string sync_interval = 4;
}

enum ProjectSourceState {
  Pending = 0;
  Synchronising = 1;
  Synchronized = 2;
  Error = 3;
}

enum ProjectSourceIssueType {
  Unknown = 0;
  AuthFailure = 1;
  RepositoryNotFound = 2;
  AmeProjectNotFound = 3;
  GitSecretNotFound = 4;
}

message ProjectSourceStatus {
  optional string last_synced = 1;
  ProjectSourceState state = 2;
  optional string reason = 3;
  repeated ProjectSourceIssue issues = 4;
}

message ProjectSourceIssue {
  ProjectSourceIssueType issue_type = 1;
  optional string explanation = 2;
}

message ProjectSourceId {
  string name = 1;
}

message TrainRequest {
  string projectid = 1;
  string model_name = 2;
}

// This is a secret stored by AME
message AmeSecret {
  string key = 1;
  string value = 2;
}

message AmeSecretId {
  string key = 1;
}

message AmeSecrets {
repeated AmeSecretId secrets = 1;
}

message ResourceCfg {
  oneof cfg {
    ProjectSourceCfg projectSrcCfg = 1;
  }
}

message ResourceId {
  oneof id {
    ProjectSourceId projectSrcId = 1;
  }
}


message ProjectSourceListParams {}

message ResourceListParams {
  oneof params {
    ProjectSourceListParams projectSourceListParams = 1;
  }
}

message ResourceIds {
  repeated ResourceId ids = 1;
}

message ListProjectSrcsResponse {
  repeated ProjectSourceCfg cfgs = 2;
}

message ProjectSrcPatchRequest {
  ProjectSourceId id = 1;
  ProjectSourceCfg cfg = 2;
}

message ProjectSrcIdRequest {
  string repo = 1;
}

message TaskRef {
  string name = 1;
  optional string project = 2;
}

message TaskCfg {
  optional string name = 1;
  optional TaskRef taskRef = 2;  
  map<string, string> resources = 3;
  oneof executor {
    PoetryExecutor poetry = 4; 
    MlflowExecutor mlflow = 5;
    PipEnvExecutor pipEnv = 6;
    PipExecutor pip = 8;
    CustomExecutor custom = 9;
  }
  repeated string dataSets = 7;
  optional TemplateRef fromTemplate = 10;
  optional ArtifactCfg artifactCfg = 11;
  optional TriggerCfg triggers = 12;
  repeated EnvVar env = 13;
  repeated Secret secrets = 14;
}

message Secret {
  oneof variant {
    AmeSecretVariant ame = 1;
  }
}

message AmeSecretVariant {
  string key = 1;
  string injectAs = 2;
}

message EnvVar {
  string key = 1;
  string val = 2;
}

message TriggerCfg {
  optional string schedule = 1;
}

message ArtifactCfg {
  bool save_changed_files = 1;
  repeated string paths = 2; 
}

message TemplateRef {
  string name = 1;
  optional string project = 2;
}

message CustomExecutor {
  string pythonVersion = 1;
  string command = 2;
}

message PoetryExecutor {
  string pythonVersion = 1;  
  string command = 2;
}


message PipExecutor {
  string pythonVersion = 1;  
  string command = 2;
}

message MlflowExecutor {
  
}

message PipEnvExecutor {
  string command = 1;
}

// TODO: should there be an error case?
message TaskStatus {
  oneof phase {
    TaskPhasePending pending = 2;  
    TaskPhaseRunning running = 3;
    TaskPhaseFailed failed = 4;
    TaskPhaseSucceeded succeeded= 5;
  }
}

message TaskPhasePending {
  
}

message TaskPhaseRunning {
  string workflowName= 1;
}
message TaskPhaseFailed {
  string workflowName= 1;
}
message TaskPhaseSucceeded {
  string workflowName= 1;
}

message DataSetCfg {
 string name = 1;
 TaskCfg  task = 2;
 string path = 3;
 optional string size = 4;
}

message ProjectCfg {
  string name = 1;
  repeated Model models = 2;
  repeated DataSetCfg dataSets = 3;
  repeated TaskCfg tasks = 4;
  repeated TaskCfg templates = 5;
  optional bool enableTriggers = 6;
}

message ProjectStatus {
  map<string,ModelStatus> models = 1;
}

message ModelStatus {
  optional string latestValidatedModelVersion = 1;
}

message CreateProjectRequest {
  ProjectCfg cfg = 1;
  optional bool enableTriggers = 2;
}

message ProjectId {
  string name = 1;
}

message Model {
  string name = 1;
  optional TaskCfg validationTask = 2; 
  optional ModelTrainingCfg training = 3;
  optional ModelDeploymentCfg deployment = 4;
}

message ModelTrainingCfg {
  TaskCfg task = 1;
}

message ModelDeploymentCfg {
   map<string, string> ingressAnnotations = 1;
   optional int32 replicas = 2;
   optional string image = 3;
   map<string,string> resources = 4;
   optional bool enableTls = 5;
}

message RunTaskRequest {
  ProjectId projectId= 1;
  TaskCfg taskCfg = 2;
}

message TaskId {
  string name = 1;
}

message ListTasksRequest {
  
}

message ListTasksResponse {
  map<string, TaskListEntry> tasks = 1;
}

message TaskListEntry {
  TaskStatus status = 1;
  string timeStamp = 2;
}

message RemoveTaskRequest {
  string name = 1;
  optional bool approve = 2;
}

service AmeService {
  rpc RunTask(RunTaskRequest) returns (TaskIdentifier) {}
  rpc GetTask(TaskIdentifier) returns (TaskCfg) {}
  rpc DeleteTask(TaskIdentifier) returns (Empty) {}
  rpc CreateTaskProjectDirectory(TaskProjectDirectoryStructure) returns (Empty) {}
  rpc UploadProjectFile(stream ProjectFileChunk) returns (Empty) {}
  rpc StreamTaskLogs(TaskLogRequest) returns (stream LogEntry) {}
  rpc CreateProjectSrc(ProjectSourceCfg) returns (ProjectSourceId) {}
  rpc DeleteProjectSrc(ProjectSourceId) returns (Empty) {}
  rpc TrainModel(TrainRequest) returns (Empty) {}
  rpc WatchProjectSrc(ProjectSourceId) returns (stream ProjectSourceStatus) {}
  rpc CreateSecret(AmeSecret) returns (Empty) {}
  rpc DeleteSecret(AmeSecretId) returns (Empty) {}
  rpc ListSecrets(Empty) returns (AmeSecrets) {} 
  rpc UpdateProjectSrc(ProjectSrcPatchRequest) returns (Empty) {}
  rpc CreateResource(ResourceCfg) returns (ResourceId) {}
  rpc ListResource(ResourceListParams) returns (ResourceIds) {}
  rpc GetProjectSrcCfg(ProjectSourceId) returns (ProjectSourceCfg) {}
  rpc GetProjectSrcStatus(ProjectSourceId) returns (ProjectSourceStatus) {}
  rpc GetProjectSrcId(ProjectSrcIdRequest) returns (ProjectSourceId) {}
  rpc ListProjectSrcs(ProjectSourceListParams) returns (ListProjectSrcsResponse) {} 
  rpc CreateProject(CreateProjectRequest) returns (ProjectId) {}
  rpc ListTasks(ListTasksRequest) returns (ListTasksResponse) {}
  rpc RemoveTask(RemoveTaskRequest) returns (Empty) {}
 }
