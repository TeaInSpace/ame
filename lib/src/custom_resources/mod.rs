use crate::error::AmeError;
use crate::grpc::*;
use k8s_openapi::chrono::OutOfRangeError;
use k8s_openapi::chrono::ParseError;
use kube::core::ObjectMeta;
use kube::ResourceExt;
use secrets::SecretError;
use std::env::VarError;
use task::Task;
use thiserror::Error;

use self::project_source::ProjectSource;
use self::project_source::ProjectSourceSpec;

use self::task::TaskSpec;
use self::task::TaskType;

pub mod argo;
pub mod common;
pub mod data_set;
pub mod project;
pub mod project_source;
pub mod project_source_ctrl;
pub mod secrets;
pub mod task;

#[derive(Error, Debug)]
pub enum Error {
    #[error("Finalizer Error: {0}")]
    FinalizerError(#[source] kube::runtime::finalizer::Error<kube::Error>),

    #[error("SerializationError: {0}")]
    SerializationError(#[from] serde_json::Error),

    #[error("Received error from kube API: {0}")]
    KubeApiError(#[from] kube::Error),

    #[error("Failed to find projet source: {0}")]
    MissingProjectSrc(String),

    #[error("libgit2 produced an error: {0}")]
    GitError(#[from] git2::Error),

    #[error("Ame errored: {0}")]
    SerdeYamlError(#[from] serde_yaml::Error),

    #[error("Ame errored: {0}")]
    FileError(#[from] std::io::Error),

    #[error("Ame erroed: {0}")]
    SystemTimeError(#[from] std::time::SystemTimeError),

    #[error("Ame errored: {0}")]
    K8sTimeError(#[from] OutOfRangeError),

    #[error("Invalid project source: {0}")]
    InvalidProjectSrc(String),

    #[error("Environment variable was not present: {0}")]
    MissingEnvVariable(#[from] VarError),

    #[error("No model deployment was found")]
    MissingDeployment(),

    #[error("No matching template found was found: {0} {1}")]
    MissingTemplate(String, String),

    #[error("No Mlfow URL was found")]
    MissingMlflowUrl(),

    #[error("Request error: {0}")]
    RequestError(#[from] reqwest::Error),

    #[error("Failed to merge structs: {0}")]
    MergeError(#[from] serde_merge::error::Error),

    #[error("failed to pass duration: {0}")]
    InvalidDuration(String),

    #[error("got error while converting: {0}")]
    ConversionError(String),

    #[error("got error while passing time: {0}")]
    ChronoParseError(#[from] ParseError),

    #[error("got error from AME's secret store: {0}")]
    SecretError(#[from] SecretError),

    #[error("failed to find model status for: {0}")]
    MissingModelStatus(String),

    #[error("failed to find validation task status for model: {0}")]
    MissingValidationTask(String),

    #[error("failed to find project id for task: {0}")]
    MissingProjectId(String),

    #[error("failed to find project with id: {0}")]
    MissingProject(String),

    #[error("failed to find data for task with name : {0}")]
    MissingDataSets(String),

    #[error("failed to find AME file project source with name : {0}")]
    MissingAmeFile(String),
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

impl TryFrom<self::CreateTaskRequest> for Task {
    type Error = Error;

    fn try_from(t: CreateTaskRequest) -> Result<Self> {
        let CreateTaskRequest {
            id: Some(TaskIdentifier { name: id }),
            template: Some(template),
        } = t else {
            return Err(Error::ConversionError("Failed to extract id and template from CreateTaskRequest".to_string()))
        };

        Ok(Task {
            metadata: ObjectMeta {
                name: Some(id),
                ..ObjectMeta::default()
            },
            spec: TaskSpec {
                projectid: Some(template.projectid),
                runcommand: Some(template.command),
                image: template.image,
                task_type: template.task_type.map(|t| match t {
                    2 => TaskType::Poetry,
                    1 => TaskType::Mlflow,
                    _ => TaskType::PipEnv,
                }),
                ..TaskSpec::default()
            },
            status: None,
        })
    }
}

impl From<ProjectSourceCfg> for ProjectSource {
    fn from(project_src: ProjectSourceCfg) -> Self {
        ProjectSource {
            metadata: ObjectMeta {
                generate_name: Some("ameprojectsrc".to_string()),
                ..ObjectMeta::default()
            },
            spec: ProjectSourceSpec { cfg: project_src },
            status: None,
        }
    }
}

impl From<TaskTemplate> for Task {
    fn from(t: TaskTemplate) -> Self {
        Task {
            metadata: ObjectMeta {
                generate_name: Some("mytask".to_string()),
                ..ObjectMeta::default()
            },
            spec: TaskSpec {
                projectid: Some(t.projectid),
                runcommand: Some(t.command),
                image: t.image,
                task_type: t.task_type.map(|t| {
                    if t == 1 {
                        TaskType::Mlflow
                    } else {
                        TaskType::PipEnv
                    }
                }),
                ..TaskSpec::default()
            },
            status: None,
        }
    }
}

impl From<Task> for TaskTemplate {
    fn from(t: Task) -> Self {
        TaskTemplate {
            name: "".to_string(),
            command: t.spec.runcommand.unwrap_or("".to_string()),
            projectid: t.spec.projectid.unwrap_or("".to_string()),
            image: t.spec.image,
            task_type: t.spec.task_type.map(|t| match t {
                TaskType::Mlflow => 1,
                TaskType::Poetry => 2,
                TaskType::PipEnv => 0,
            }),
        }
    }
}

impl From<Task> for TaskIdentifier {
    fn from(t: Task) -> Self {
        TaskIdentifier { name: t.name_any() }
    }
}

impl TryFrom<TaskCfg> for TaskSpec {
    type Error = AmeError;
    fn try_from(cfg: TaskCfg) -> crate::Result<Self> {
        Ok(TaskSpec::from_ref(cfg.task_ref.ok_or(
            AmeError::MissingTaskRef(cfg.name.unwrap_or_default()),
        )?))
    }
}
