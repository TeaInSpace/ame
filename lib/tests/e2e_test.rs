use ame::custom_resources::{
    data_set::{DataSet, DataSetPhase, DataSetStatus},
    new_task::Task,
};

use ame::{
    custom_resources::{
        common::setup_cluster, project::Project, project_source_ctrl::ProjectSrcCtrl,
        secrets::SecretCtrl,
    },
    grpc::{task_status::Phase, ProjectCfg, TaskCfg, TaskStatus},
};
use assert_cmd::prelude::*;
use fs_extra::dir::CopyOptions;

use futures_util::StreamExt;

use k8s_openapi::api::{apps::v1::Deployment, core::v1::Service, networking::v1::Ingress};
use kube::{
    api::{DeleteParams, ListParams, PatchParams},
    runtime::{watcher, WatchStreamExt},
    Api, Client,
};
use rstest::*;
use serial_test::serial;
use std::collections::HashMap;

use time::Instant;

use std::{
    path::{Path, PathBuf},
    process::Command,
};
use tracing::{debug, instrument};
use tracing_subscriber::EnvFilter;

static INGRESS_NAMESPACE: &str = "ingress-nginx";
static INGRESS_SERVICE: &str = "ingress-nginx-controller";
static AME_NAMESPACE: &str = "ame-system";

async fn test_setup() -> Result<(), Box<dyn std::error::Error>> {
    // std::env::set_var(
    //     "AME_ENDPOINT",
    //     find_ame_endpoint(INGRESS_NAMESPACE, INGRESS_SERVICE).await?,
    // );
    std::env::set_var("AME_ENDPOINT", "http://localhost:3342");

    Ok(())
}

fn prepare_test_project(path_from_root: &str) -> Result<PathBuf, Box<dyn std::error::Error>> {
    let mut test_project_path = PathBuf::from(path_from_root);

    for i in 0..10 {
        if test_project_path.is_dir() && i < 9 {
            break;
        }

        if i < 9 {
            test_project_path = Path::new("..").join(test_project_path);
            continue;
        }

        return Err(format!(
            "failed to find test project directory: {}",
            test_project_path.display()
        ))?;
    }

    let temp_dir = assert_fs::TempDir::new()?.into_persistent();

    fs_extra::copy_items(
        &[test_project_path],
        temp_dir.path(),
        &CopyOptions::default(),
    )?;

    Ok(temp_dir
        .path()
        .join(Path::new(path_from_root).file_name().unwrap()))
}

// TODO add concurrent tests
// TODO Test failure handling

#[rstest]
#[trace]
#[case("test_data/test_projects/executors/poetry", "training")]
#[case("test_data/test_projects/executors/pipenv", "training")]
#[case("test_data/test_projects/executors/pip", "training")]
#[case("test_data/test_projects/executors/mlflow", "training")]
#[case(
    "test_data/test_projects/executors/mlflow",
    "training-with-local-template"
)]
#[case(
    "test_data/test_projects/executors/mlflow",
    "training-with-remote-template"
)]
#[case("test_data/test_projects/executors/custom", "training")]
#[case("test_data/test_projects/env", "testenv")]
#[case("test_data/test_projects/env", "testsecret")]
#[ignore]
#[tokio::test]
async fn ame_run_task(
    #[case] project_dir: &str,
    #[case] task_id: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    init_test_tracing();
    let _ = setup_cluster("ame-system").await; // TODO: async tests are affecting each other here.
    debug!("running task {task_id} in {project_dir}");
    let projects: Api<Project> = Api::namespaced(kube_client().await?, AME_NAMESPACE);
    let project: Project = serde_json::from_value(serde_json::json!(
        {
            "metadata": {
            "name": "shared-templates-project"
                },
            "spec": {
                "name": "shared-templates",
                "templates": [{

                "name": "mlflow-template",
                "executor": {
                    "mlflow": {}
                }}],

                "deletionApproved": false
            }
        }
    ))
    .unwrap();

    projects
        .patch(
            "shared-templates-project",
            &PatchParams::apply("AME-TEST").force(),
            &kube::api::Patch::Apply(project),
        )
        .await
        .unwrap();

    let temp = prepare_test_project(project_dir)?;
    let mut cmd = Command::cargo_bin("ame")?;
    test_setup().await?;

    let res = cmd
        .current_dir(temp.clone())
        .arg("task")
        .arg("run")
        .arg("--logs")
        .arg(task_id)
        .assert()
        .success();

    let mut settings = insta::Settings::clone_current();
    settings.add_filter("time=\".*\".*\n", "");
    settings.add_filter(
        "created virtual environment .*",
        "created virtual environment \"redacted\"",
    );
    settings.add_filter("creator .*", "creator \"redacted\"");
    settings.add_filter(
        "\\d\\d\\d\\d/\\d\\d/\\d\\d \\d\\d:\\d\\d:\\d\\d",
        "\"redacted timestamp\"",
    );
    settings.add_filter("run .*", "\"redacted run ID\"");
    settings.add_filter("ID '.*'", "\"redacted run ID\"");
    settings.add_filter("tmp/tmp.*\\s", "\"redacted temporary directory\"");
    settings.add_filter("mlflow-.*/", "\"redacted MLflow env ID\"");
    settings.add_filter(".*: UserWarning: .*\\n", "");
    settings.add_filter("warnings\\.warn.*\\n", "");
    settings.add_filter("  \"redacted timestamp", "\"redacted timestamp");
    settings.add_filter("  Score:", "Score:");
    settings.add_filter("added seed packages.*", "redacted");
    settings.add_filter("Registered model '.*' already exists.*", "redacted");
    settings.add_filter("Created version '.'.*\\n", "");
    settings.add_filter(", version .", ", version %");
    settings.add_filter("Successfully registered model .*", "redacted");
    // TODO: why does \\. not work here?
    settings.add_filter("\\d+\\.\\d+ MB", "redacted");
    settings.add_filter("\\d+\\.\\d+ KB", "redacted");
    settings.add_filter("tasks/.+/projectfiles", "tasks/redacted/projectfiles");
    let _guard = settings.bind_to_scope();

    // TODO: how can we incorporate snapshots in a non brittle way?
    // insta::assert_snapshot!(&String::from_utf8(res.get_output().stdout.clone())?);

    let project = ProjectCfg::try_from_dir(temp.to_str().unwrap()).unwrap();

    debug!(
        "stdout: {}",
        String::from_utf8(res.get_output().stdout.clone()).unwrap()
    );

    // TODO: should be use the ame API server to monitor the status here?
    let tasks: Api<Task> = Api::namespaced(kube_client().await?.clone(), AME_NAMESPACE);
    let task_list = tasks.list(&ListParams::default()).await?;
    let task_obs: Vec<Task> = task_list
        .items
        .into_iter()
        .filter(|t| {
            t.spec.cfg.name.as_ref().unwrap() == task_id
                && t.metadata
                    .owner_references
                    .as_ref()
                    .unwrap()
                    .iter()
                    .any(|oref| oref.name.contains(&project.name))
        })
        .collect();

    assert_eq!(
        task_obs.len(),
        1,
        "expected to find a single matching task, instead found {}",
        task_obs.len()
    );

    // TODO: we need to identify each task uniqely here
    assert!(
        task_obs[0]
            .status
            .as_ref()
            .unwrap()
            .phase
            .as_ref()
            .unwrap()
            .success(),
        "task {} was not successful",
        { task_id }
    );

    Ok(())
}
#[rstest]
#[trace]
#[case("test_data/test_projects/executors/poetry", "crontraining")]
#[ignore]
#[tokio::test]
#[serial]
async fn ame_trigger_tasks(
    #[case] project_dir: &str,
    #[case] task_id: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    init_test_tracing();
    let _ = setup_cluster("ame-system").await; // TODO: async tests are affecting each other here.
    let _projects: Api<Project> = Api::namespaced(kube_client().await?, AME_NAMESPACE);
    let tasks: Api<Task> = Api::namespaced(kube_client().await?, AME_NAMESPACE);

    let temp = prepare_test_project(project_dir)?;
    let mut cmd = Command::cargo_bin("ame")?;
    test_setup().await?;

    let mut project = ProjectCfg::try_from_dir(temp.to_str().unwrap()).unwrap();

    let now = time::OffsetDateTime::now_utc();

    let target = now.checked_add(time::Duration::seconds(270)).unwrap();

    let mut project_file_path = temp.clone();
    project_file_path.push("ame.yaml");

    project.tasks[0].triggers = Some(ame::grpc::TriggerCfg {
        schedule: Some(format!("{} {} * * *", target.minute(), target.hour())),
    });

    let f = std::fs::File::create(project_file_path)?;
    serde_yaml::to_writer(f, &project).unwrap();

    let res = cmd
        .current_dir(temp.clone())
        .arg("project")
        .arg("push")
        .arg("--triggers")
        .assert()
        .success();

    let start_t = Instant::now();

    loop {
        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
        if start_t.elapsed() > time::Duration::seconds(360) {
            return Err("task took too long to complete".to_string().into());
        }
        let task_list = tasks.list(&ListParams::default()).await?;
        let task_obs: Vec<Task> = task_list
            .items
            .into_iter()
            .filter(|t| {
                t.spec.cfg.name.as_ref().unwrap() == task_id
                    && t.metadata
                        .owner_references
                        .as_ref()
                        .unwrap()
                        .iter()
                        .any(|oref| oref.name.contains(&project.name))
            })
            .collect();

        assert!(
            task_obs.len() < 2,
            "expected to find a single or no matching task, instead found {}",
            task_obs.len(),
        );

        let Some(task) = task_obs.get(0) else {
            continue;
        };

        let _now = time::OffsetDateTime::now_utc();

        if start_t.elapsed() < time::Duration::seconds(120) {
            return Err(format!(
                "we do not expect to see the schedule task a head of time {}",
                start_t.elapsed()
            )
            .to_string()
            .into());
        }

        if task
            .status
            .as_ref()
            .unwrap()
            .phase
            .as_ref()
            .unwrap()
            .success()
        {
            break;
        }
    }

    debug!(
        "stdout: {}",
        String::from_utf8(res.get_output().stdout.clone()).unwrap()
    );

    Ok(())
}

// TODO get rid of this, we don't need a fixture for something so simple
#[fixture]
async fn kube_client() -> Result<Client, kube::Error> {
    Client::try_default().await
}

pub async fn delete_registered_model(name: &str, mlflow_url: &str) -> Result<(), reqwest::Error> {
    let mut map = HashMap::new();
    map.insert("name", &name);

    let res = reqwest::Client::new()
        .delete(format!(
            "{mlflow_url}/api/2.0/mlflow/registered-models/delete"
        ))
        .json(&map);

    res.send().await?.error_for_status()?;

    Ok(())
}

fn init_test_tracing() {
    let _ = tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .try_init();
}

#[tokio::test]
#[serial]
#[ignore]
#[instrument]
async fn can_use_data_set_train_validate_and_deploy_model() -> Result<(), Box<dyn std::error::Error>>
{
    debug!("can use data, preparing cluster");
    init_test_tracing();
    let _ = setup_cluster(AME_NAMESPACE).await;
    test_setup().await?;

    debug!("Cluster is ready!");

    // The template repo is required as the ame-demo requires it to train.
    let template_repo = "https://github.com/TeaInSpace/ame-template-demo.git";
    let data_set_repo = "https://github.com/TeaInSpace/ame-dataset-demo.git";
    let project_repo = "https://github.com/TeaInSpace/ame-demo.git";
    let model_name = "logreg"; // this name is from the ame-demo repo.
    let project_src_ctrl = ProjectSrcCtrl::new(kube_client().await?, AME_NAMESPACE);

    let kube_client = kube_client().await?;
    let deployments: Api<Deployment> = Api::namespaced(kube_client.clone(), AME_NAMESPACE);
    let datasets: Api<DataSet> = Api::namespaced(kube_client.clone(), AME_NAMESPACE);
    let projects: Api<Project> = Api::namespaced(kube_client.clone(), AME_NAMESPACE);
    let ingress: Api<Ingress> = Api::namespaced(kube_client.clone(), AME_NAMESPACE);
    let services: Api<Service> = Api::namespaced(kube_client.clone(), AME_NAMESPACE);
    let tasks: Api<Task> = Api::namespaced(kube_client.clone(), AME_NAMESPACE);

    // TODO: whyid this not throw an error with AME_FILE_NAME.
    let secret_ctrl = SecretCtrl::try_default(AME_NAMESPACE).await?;
    let s3_secret = "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG";

    secret_ctrl
        .store_secret_if_empty("s3secretkey", s3_secret.to_string())
        .await?;

    let res = delete_registered_model("logreg", "http://localhost:5000").await;

    // If a model was not present mlflow should return a 404. In this case we can
    // proceed as that does not indicate a problem. Any other error code indicates
    // that something went wrong and the test therefor aborts.
    if let Err(e) = res {
        if let Some(status) = e.status() {
            if status != 404 {
                return Err(format!("received unexpected error from mlflow server {e}").into());
            }
        }
    }

    // TODO: this is very brittle, should be replaced by proper resource ownership.
    let _ = deployments.delete("logreg", &DeleteParams::default()).await;
    let _ = ingress.delete("logreg", &DeleteParams::default()).await;
    let _ = services.delete("logreg", &DeleteParams::default()).await;

    // TODO: test when controller is blocked from reconciling due to a deletion request.

    // Create each project src using the CLI.
    // This should be the only interaction necessary to deploy a model.
    debug!("create project sources with the CLI");
    for repo in [template_repo, data_set_repo, project_repo] {
        let mut cmd = Command::cargo_bin("ame")?;
        let _output = cmd
            .arg("projectsrc")
            .arg("create")
            .arg(repo)
            .assert()
            .success();
    }

    // It is important that we only have 3 projects as the watcher
    // does not perform any filtering. If other projects are present
    // something is interfering with the test.
    assert_eq!(
        projects
            .list_metadata(&ListParams::default())
            .await?
            .items
            .len(),
        3
    );

    // No deployment for the model should be present.
    // TODO: this should be replaced by ownershop somehow.
    assert!(deployments.get(model_name).await.is_err());

    let timeout = std::time::Duration::from_secs(300);
    let start = std::time::Instant::now();

    let mut data_set_watcher = watcher(datasets, ListParams::default())
        .applied_objects()
        .boxed();

    debug!("waiting for dataset to be ready");
    // Before the model can be trained we expect a data set to be prepared.
    while let Some(data_set) = data_set_watcher.next().await {
        if let DataSet {
            status:
                Some(DataSetStatus {
                    phase: Some(DataSetPhase::Ready { .. }),
                }),
            ..
        } = data_set?
        {
            break;
        }

        if start.elapsed() > timeout {
            return Err("failed to prepare dataset within timeout, with event"
                .to_string()
                .into());
        }
    }

    let mut task_watcher = watcher(tasks, ListParams::default())
        .applied_objects()
        .boxed();

    // Before the model can be deployed we expect it to be trained and then validated.
    for target_name in ["training", "mlflow_validation"] {
        let start = std::time::Instant::now();
        debug!("waiting for task {target_name}");
        while let Some(e) = task_watcher.next().await {
            let task = e?;

            let TaskCfg {
                name: Some(task_name),
                ..
            } = task.spec.cfg
            else {
                return Err(format!("{target_name} task was missing a name in the cfg").into());
            };

            if task_name != target_name {
                continue;
            }

            if let Some(
                TaskStatus {
                    phase: Some(Phase::Succeeded(_)),
                },
                ..,
            ) = task.status
            {
                break;
            }

            if start.elapsed() > timeout {
                return Err(format!("failed to start task {} within timeout", task_name).into());
            }
        }
    }

    let mut deployment_watcher = watcher(
        deployments,
        ListParams::default().fields(&format!("metadata.name=={model_name}")),
    )
    .applied_objects()
    .boxed();

    debug!("waiting for deployment");
    let start = std::time::Instant::now();
    while let Some(e) = deployment_watcher.next().await {
        let Some(status) = e?.status else {
            return Err("missing deployment status".into());
        };

        if let Some(1) = status.ready_replicas {
            break;
        }

        if start.elapsed() > timeout {
            return Err("failed to deploy model within timeout".into());
        }
    }

    project_src_ctrl
        .delete_project_src_for_repo(project_repo)
        .await?;
    project_src_ctrl
        .delete_project_src_for_repo(data_set_repo)
        .await?;
    project_src_ctrl
        .delete_project_src_for_repo(template_repo)
        .await?;

    Ok(())
}
