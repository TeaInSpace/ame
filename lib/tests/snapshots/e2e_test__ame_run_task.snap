---
source: lib/tests/e2e_test.rs
expression: "&String::from_utf8(res.get_output().stdout.clone())?"
---
Uploading project!
download: 's3://ame/tasks/redacted/projectfiles/ame.yaml' -> './ame.yaml' (149 bytes in 0.0 seconds, redacted/s)
download: 's3://ame/tasks/redacted/projectfiles/nn.py' -> './nn.py' (58 bytes in 0.0 seconds, redacted/s)
download: 's3://ame/tasks/redacted/projectfiles/poetry.lock' -> './poetry.lock' (4587 bytes in 0.0 seconds, redacted/s)
download: 's3://ame/tasks/redacted/projectfiles/pyproject.toml' -> './pyproject.toml' (297 bytes in 0.0 seconds, redacted/s)
Installing Python-3.10.9...
Installed Python-3.10.9 to /home/ame/.pyenv/versions/3.10.9
The currently activated Python version 3.11.4 is not supported by the project (3.10.9).
Trying to find and use a compatible version. 
Using python3 (3.10.9)
Creating virtualenv test-6gE1vKXj-py3.10 in /home/ame/.cache/pypoetry/virtualenvs
Installing dependencies from lock file

Package operations: 1 install, 0 updates, 0 removals

  â€¢ Installing numpy (1.24.3)
The currently activated Python version 3.11.4 is not supported by the project (3.10.9).
Trying to find and use a compatible version. 
Using python3 (3.10.9)
np array:  [1 2 3]
WARNING: Exiting now because of --dry-run
Uploading artifacts: './ame.yaml'
'./nn.py'
'./poetry.lock'
'./pyproject.toml'
WARNING: Module python-magic is not available. Guessing MIME types based on file extensions.
upload: 'ame.yaml' -> 's3://ame/tasks/trainingclzjj/artifacts/ame.yaml' (149 bytes in 0.0 seconds, redacted/s) [1 of 1]
WARNING: Module python-magic is not available. Guessing MIME types based on file extensions.
upload: 'nn.py' -> 's3://ame/tasks/trainingclzjj/artifacts/nn.py' (58 bytes in 0.0 seconds, redacted/s) [1 of 1]
WARNING: Module python-magic is not available. Guessing MIME types based on file extensions.
upload: 'poetry.lock' -> 's3://ame/tasks/trainingclzjj/artifacts/poetry.lock' (4587 bytes in 0.0 seconds, redacted/s) [1 of 1]
WARNING: Module python-magic is not available. Guessing MIME types based on file extensions.
upload: 'pyproject.toml' -> 's3://ame/tasks/trainingclzjj/artifacts/pyproject.toml' (297 bytes in 0.0 seconds, redacted/s) [1 of 1]

